plugins/pulp_docker/plugins/token_util.py

```diff
@@ -0,0 +1,82 @@
+from cStringIO import StringIO
+import json
+import logging
+import urllib
+
+from nectar.request import DownloadRequest
+
+
+_logger = logging.getLogger(__name__)
+
+
+def add_auth_header(request, token):
+    """
+    Adds the token into the request's headers as specified in the Docker v2 API documentation.
+
+    https://docs.docker.com/registry/spec/auth/token/#using-the-bearer-token
+
+    :param request: a download request
+    :type  request: nectar.request.DownloadRequest
+    """
+    if request.headers is None:
+        request.headers = {}
+    request.headers['Authorization'] = 'Bearer %s' % token
+
+
+def request_token(downloader, request, response_headers):
+    """
+    Attempts to retrieve the correct token based on the 401 response header.
+
+    According to the Docker API v2 documentation, the token be retrieved by issuing a GET
+    request to the url specified by the `realm` within the `WWW-Authenticate` header. This
+    request should add the following query parameters:
+
+        service: the name of the service that hosts the desired resource
+        scope:   the specific resource and permissions requested
+
+    https://docs.docker.com/registry/spec/auth/token/#requesting-a-token
+
+    :param request: a download request
+    :type  request: nectar.request.DownloadRequest
+    :param response_headers: headers from the 401 response
+    :type  response_headers: basestring
+    :return: Bearer token for requested resource
+    :rtype:  str
+    """
+    auth_info = parse_401_response_headers(response_headers)
+    try:
+        token_url = auth_info.pop('realm')
+    except KeyError:
+        raise IOError("No realm specified for token auth challenge.")
+
+    token_url = "{url}?{query}".format(url=token_url, query=urllib.urlencode(auth_info))
```

Comment:
In case the url already has one or more query parameters, or a fragment, the right way to do this is ``urlparse`` it, add your query parameter to the result, then ``urlunparse`` it.